src/mrinufft/_utils.py

```diff
@@ -0,0 +1,82 @@
+"""General utility functions for MRI-NUFFT."""
+
+import warnings
+from collections import defaultdict
+from functools import wraps
+import numpy as np
+
+
+def proper_trajectory(trajectory, normalize="pi"):
+    """Normalize the trajectory to be used by NUFFT operators.
+
+    Parameters
+    ----------
+    trajectory: np.ndarray
+        The trajectory to normalize, it might be of shape (Nc, Ns, dim) of (Ns, dim)
+
+    normalize: str
+        if "pi" trajectory will be rescaled in [-pi, pi], if it was in [-0.5, 0.5]
+        if "unit" trajectory will be rescaled in [-0.5, 0.5] if it was not [-0.5, 0.5]
+
+    Returns
+    -------
+    new_traj: np.ndarray
+        The normalized trajectory of shape (Nc * Ns, dim) or (Ns, dim) in -pi, pi
+    """
+    # flatten to a list of point
+    try:
+        new_traj = np.asarray(trajectory).copy()
+    except Exception as e:
+        raise ValueError(
+            "trajectory should be array_like, with the last dimension being coordinates"
+        ) from e
+    new_traj = new_traj.reshape(-1, trajectory.shape[-1])
+
+    if normalize == "pi" and np.max(abs(new_traj)) - 1e-4 < 0.5:
+        warnings.warn(
+            "Samples will be rescaled to [-pi, pi), assuming they were in [-0.5, 0.5)"
+        )
+        new_traj *= 2 * np.pi
+    elif normalize == "unit" and np.max(abs(new_traj)) - 1e-4 > 0.5:
+        warnings.warn(
+            "Samples will be rescaled to [-0.5, 0.5), assuming they were in [-pi, pi)"
+        )
+        new_traj /= 2 * np.pi
+    if normalize == "unit" and np.max(new_traj) >= 0.5:
+        new_traj = (new_traj + 0.5) % 1 - 0.5
+    return new_traj
+
+
+class MethodRegister:
```

Comment:
Was this abstraction really needed and helpful? Especially when we have only 3 density methods and also in which only 2 are used. Although, i understand you are trying to be future proof.

Reply:
>  Although, i understand you are trying to be future proof.
Yep, Maybe we would like to collect other type of methods as well at one point (e.g. trajectories)